feat: adds dedicated Testing section to Tools

dutterbutter opened the pull request:
**Note:** This will remain a DRAFT PR until we are done ensuring `era-test-node` works as intended and the documentation introduced in this PR is updated with any changes done. 

# What 💻

- Adds dedicated section for "Local Testing" in Tools.
- Offers an overview of various local testing alternatives available for developers, outlining when and how to utilize each, as well as their distinct functionalities.
- Splits out Dockerized local testing into own page 
- Splits out In-memory node local testing into own page 

# Why ✋

- With the addition of [`era-test-node`](https://github.com/matter-labs/era-test-node) which provides developers with forking capabilities and additional features, providing clear and concise guidance on the use and implementation of these testing options has become essential. This update aims to aid developers with instructions and examples for effective usage.

# Evidence 📷 

- Run: `yarn dev:docs` and visit `http://localhost:8080/docs/`

![Screenshot 2023-07-07 at 12 02 01 PM](https://github.com/matter-labs/zksync-web-era-docs/assets/29983536/8d53b136-c09d-4ba0-8f6d-7304e5b7edd2)

uF4No commented:
Let's also add the network details in each page (docker and in-memory). Something like:

```md
### Network details
L1 RPC: http://localhost:8545
L2 RPC: http://localhost:3050
Network Id: 270
```

uF4No commented:
@dutterbutter missing "Prerequisites" in the in-memory node page?

uF4No commented on docs/tools/local-testing/dockerized-testing.md:
```diff
@@ -0,0 +1,187 @@
+# Dockerized testing
```
```suggestion
# Dockerized setup
```
Changing text for menu

uF4No commented on docs/tools/local-testing/era-test-node.md:
```diff
@@ -0,0 +1,362 @@
+# In-Memory node testing
```
```suggestion
# In-memory node
```
changing text for menu

idea404 reviewed (commented):
LGTM. I suggest updating the examples such that the key differences between both testing methods are demonstrated in each.

dutterbutter commented:
@uF4No @idea404 @MexicanAce opening this up for an additional review. I have added the requested changes, and added essential warnings indicating its **alpha stage**.

uF4No commented on firebase.json:
```diff
@@ -438,6 +443,11 @@
+        {
+          "source": "/docs/tools/hardhat/testing.html",
+          "destination": "/docs/tools/local-testing/",
+          "type": "302"
```
```suggestion
          "type": "301"
```
Let's do 301 redirects as those are permanent

uF4No commented on firebase.json:
```diff
@@ -293,6 +293,11 @@
+        {
+          "source": "/docs/tools/hardhat/testing.html",
+          "destination": "/docs/tools/local-testing/",
+          "type": "302"
```
```suggestion
          "type": "301"
```

uF4No commented on docs/tools/local-testing/README.md:
```diff
@@ -0,0 +1,46 @@
+# Local testing strategies for zkSync Era
+
+zkSync Era provides two distinct testing environments for your development needs - Dockerized Local Setup and the In-Memory Node. Each solution boasts unique characteristics tailored to diverse use cases. This guide aims to unpack the intricacies of these tools, aiding you in selecting the setup best suited for your development workflow.
```
```suggestion
zkSync Era provides two distinct testing environments for your development needs 

- Dockerized Local Setup 
- In-Memory Node. 

Each solution boasts unique characteristics tailored to diverse use cases. This guide aims to unpack the intricacies of these tools, aiding you in selecting the setup best suited for your development workflow.
```

uF4No commented on docs/tools/local-testing/README.md:
```diff
@@ -0,0 +1,46 @@
+
+zkSync Era provides two distinct testing environments for your development needs - Dockerized Local Setup and the In-Memory Node. Each solution boasts unique characteristics tailored to diverse use cases. This guide aims to unpack the intricacies of these tools, aiding you in selecting the setup best suited for your development workflow.
+
+## Understanding your choices: In-Memory node vs Dockerized local setup
```
```suggestion
## In-Memory node vs Dockerized local setup
```
Just to have a shorter title in the table of contents on the side

uF4No commented on docs/tools/local-testing/README.md:
```diff
@@ -0,0 +1,46 @@
+
+1. **Dockerized local setup**: An extensive zkSync Era network simulation that comprises a Postgres database, a local Geth node functioning as Layer 1, and the zkSync node. Opt for this setup for comprehensive simulations and testing that require interaction with both L1 and L2.
+
+2. **In-Memory node**: A lightweight, speedy alternative, the In-Memory Node, supports forking the state from various networks, including the mainnet and testnet. This choice is ideal for swift testing, prototyping, and bootloader and system contract testing.
```
```suggestion
2. **In-Memory node**: A lightweight, speedy alternative, the in-memory node, supports forking the state from various networks, including the mainnet and testnet. This choice is ideal for swift testing, prototyping, and bootloader and system contract testing.
```
just a suggestion... or do we want to use capitals for it?

uF4No commented on docs/tools/local-testing/dockerized-testing.md:
```diff
@@ -0,0 +1,193 @@
+- The zkSync node itself.
+
+:::tip
+ The first `start.sh` script execution should go uninterrupted. If the bootstrapping process halts unexpectedly, reset the local zkSync state and try again.
```
```suggestion
 The first `start.sh` script execution should go uninterrupted. If the bootstrapping process halts unexpectedly, reset the local zkSync state and try again. The process can take up to 10 minutes to start (only the first time!).
```
Not sure about the phrase but we should mention it takes a while the first time

uF4No commented on docs/tools/local-testing/dockerized-testing.md:
```diff
@@ -0,0 +1,193 @@
+
+By default, the HTTP JSON-RPC API is served via port `3050`, while the WebSocket (WS) API is accessed through port `3051`.
+
+**L1 RPC:** `http://localhost:8545`
+**L2 RPC:** `http://localhost:3050`
+**Network Id:** 270
```
```suggestion
- **L1 RPC:** `http://localhost:8545`
- **L2 RPC:** `http://localhost:3050`
- **Network Id:** 270
```
They're currently displayed in a single line

uF4No commented on docs/tools/local-testing/dockerized-testing.md:
````diff
@@ -0,0 +1,193 @@
+**L2 RPC:** `http://localhost:3050`
+**Network Id:** 270
+
+### Resetting the zkSync state
+
+To reset the zkSync state, run
+
+ the `./clear.sh` script:
+
+```
+./clear.sh
+```
````
```suggestion
### Resetting the zkSync state

To reset the zkSync state, run the `./clear.sh` script:

```
./clear.sh
```
::: info Update docker images 
The `clear.sh` script will also pull the most recent Docker images so it can be used to update the dockerized local setup.
:::
```

MexicanAce commented on docs/tools/local-testing/era-test-node.md:
```diff
@@ -0,0 +1,386 @@
+Creating fork from "https://mainnet.era.zksync.io:443" L1 block: L1BatchNumber(97665) L2 block: 7989849 with timestamp 1688738510 and L1 gas price 25717302286
+Starting network with chain id: L2ChainId(270)
+Setting Rich accounts:
+Address: "0x36615Cf349d7F6344891B1e7CA7C72883F5dc049" Key: "0x7726827caac94a7f9e1b160f7ea819f172f7b6f9d2a97f992c38edeab82d4110"
```
nit: This console output is slightly out of date, but don't think it really matters

```
Creating fork from "https://testnet.era.zksync.dev:443" L1 block: L1BatchNumber(119993) L2 block: 9602073 with timestamp 1689261421 and L1 gas price 89159417930
Starting network with chain id: L2ChainId(260)

Rich Accounts
=============
Account #0: 0x36615Cf349d7F6344891B1e7CA7C72883F5dc049 (10000 ETH)
Private Key: 0x7726827caac94a7f9e1b160f7ea819f172f7b6f9d2a97f992c38edeab82d4110

Account #1: 0xa61464658AfeAf65CccaaFD3a512b69A83B77618 (10000 ETH)
Private Key: 0xac1e735be8536c6534bb4f17f06f6afc73b2b5ba84ac2cfb12f7461b20c0bbe3

Account #2: 0x0D43eB5B8a47bA8900d84AA36656c92024e9772e (10000 ETH)
Private Key: 0xd293c684d884d56f8d6abd64fc76757d3664904e309a0645baf8522ab6366d9e

Account #3: 0xA13c10C0D5bd6f79041B9835c63f91de35A15883 (10000 ETH)
Private Key: 0x850683b40d4a740aa6e745f889a6fdc8327be76e122f5aba645a5b02d0248db8

========================================
  Node is ready at 127.0.0.1:8011
========================================
```

uF4No commented on docs/tools/local-testing/README.md:
```diff
@@ -0,0 +1,46 @@
+
+### Feature comparison
+
+The following table highlights the key characteristics of each testing environment for a quick comparison:
+
+| Feature                                      | In-memory node       | Dockerized Local Setup |
+|----------------------------------------------|----------------------|------------------------|
+| Quick startup                                | Yes                  | No                     |
+| Supports forking state                       | Yes                  | No                     |
+| Console.log debugging                        | Yes                  | No                     |
+| Detailed call traces                         | Yes                  | No                     |
+| Pre-configured 'rich' accounts               | Yes                  | Yes                    |
+| Replay existing transactions                 | Yes                  | No                     |
+| Fast for integration testing                 | Yes                  | No                     |
+| Communication between Layer 1 & Layer 2      | No                   | Yes                    |
+| Multiple transactions per batch              | No                   | Yes                    |
+| Fixed values for gas estimation              | Yes                  | Varies                 |
+| Complete set of APIs                         | No (Basic set only)  | Yes                    |
```
this is awesome 🤘

uF4No commented on docs/tools/local-testing/era-test-node.md:
```diff
@@ -0,0 +1,386 @@
+
+This documentation provides instructions on setting up and using the In-Memory Node, `era-test-node`, for local testing. It covers installation, network forking, transaction details viewing, replaying transactions, and testing local bootloader and system contracts.
+
+:::warning
```
```suggestion
:::warning Tool in alpha stage
```

uF4No commented on docs/tools/local-testing/era-test-node.md:
```diff
@@ -0,0 +1,386 @@
+Before you get started, there are a few prerequisites to consider. Ensure that your development environment meets the following requirements:
+
+- **Rust:** Since `era-test-node` is written in Rust, you need to have Rust installed on your machine. You can download Rust from [here](https://www.rust-lang.org/tools/install).
+- **Other Dependencies:** This crate relies on `rocksDB` for its operation. If you encounter any compile errors due to `rocksDB`, you might also need to install `cmake`, `pkg-config`, `libssl-dev`, and `clang` using the command `apt-get install -y cmake pkg-config libssl-dev clang`.
```
```suggestion
- **Other Dependencies:** This crate relies on `rocksDB` for its operation. If you encounter any compile errors due to `rocksDB`, you might also need to install some dependencies using the following command: `apt-get install -y cmake pkg-config libssl-dev clang`.
```

uF4No commented on docs/tools/local-testing/era-test-node.md:
````diff
@@ -0,0 +1,386 @@
+```
+
+:::warning
+When utilizing `era-test-node` with MetaMask, it's essential to note that any redeployment of the in-memory node will necessitate a reset of MetaMask's cached account data. To do this, navigate to 'Settings', then 'Advanced', and finally, select 'Clear activity tab data'.
````
```suggestion
When utilizing `era-test-node` with MetaMask, it's essential to note that any restart of the in-memory node will necessitate a reset of MetaMask's cached account data (nonce, etc). To do this, navigate to 'Settings', then 'Advanced', and finally, select 'Clear activity tab data'.
```

uF4No commented on docs/tools/local-testing/era-test-node.md:
```diff
@@ -0,0 +1,386 @@
+
+### Deploying contracts
+
+Before proceeding, ensure that you've compiled your contracts using `zkforge zk-build`. For instructions on how to do this, please refer to this [link](https://github.com/matter-labs/foundry-zksync#compile-with-zkforge-zk-build).
```
Is it possible to deploy via Hardhat? If so, let's mention it first

dutterbutter replied:
You can deploy via Hardhat but the example makes use of zkforge. I will add Hardhat example as well for parity.

uF4No commented on docs/tools/testing/README.md:
```diff
@@ -0,0 +1,52 @@
+| Multiple transactions per batch         | No                  | Yes                    |
+| Fixed values for gas estimation         | Yes                 | Varies                 |
+| Complete set of APIs                    | No (Basic set only) | Yes                    |
+| wss support                             | No                  | Yes                    |
```
```suggestion
| Websocket support                             | No                  | Yes                    |
```

uF4No commented on docs/tools/testing/dockerized-testing.md:
```diff
@@ -0,0 +1,195 @@
+
+By default, the HTTP JSON-RPC API is served via port `3050`, while the WebSocket (WS) API is accessed through port `3051`.
+
+- **L1 RPC:** `http://localhost:8545`
+- **L2 RPC:** `http://localhost:3050`
+- **Network Id:** 270
```
don't you want to specify the WSS endpoints here as well?

uF4No reviewed (approved):
LGTM
